mux_client_impl/src/child.rs

```diff
@@ -0,0 +1,181 @@
+use super::{ChildStderr, ChildStdin, ChildStdout, Error, Result, Session};
+
+use core::mem::replace;
+
+use std::os::unix::process::ExitStatusExt;
+use std::process::{ExitStatus, Output};
+
+use tokio::io::AsyncReadExt;
+
+use openssh_mux_client::connection::{EstablishedSession, SessionStatus};
+
+/// Representation of a running or exited remote child process.
+///
+/// This structure is used to represent and manage remote child processes. A remote child process
+/// is created via the [`Command`](crate::Command) struct through [`Session::command`], which
+/// configures the spawning process and can itself be constructed using a builder-style interface.
+///
+/// Calling [`wait`](RemoteChild::wait) (or other functions that wrap around it) will make the
+/// parent process wait until the child has actually exited before continuing.
+///
+/// Unlike [`std::process::Child`], `RemoteChild` *does* implement [`Drop`], and will terminate the
+/// local `ssh` process corresponding to the remote process when it goes out of scope. Note that
+/// this does _not_ terminate the remote process. If you want to do that, you will need to kill it
+/// yourself by executing a remote command like `pkill` to kill it on the remote side.
+///
+/// As a result, `RemoteChild` cannot expose `stdin`, `stdout`, and `stderr` as fields for
+/// split-borrows like [`std::process::Child`] does. Instead, it exposes
+/// [`stdin`](RemoteChild::stdin), [`stdout`](RemoteChild::stdout),
+/// and [`stderr`](RemoteChild::stderr) as methods. Callers can call `.take()` to get the same
+/// effect as a split borrow and use multiple streams concurrently. Note that for the streams to be
+/// available,`Stdio::piped()` should be passed to the corresponding method on
+/// [`Command`](crate::Command).
+///
+/// ```rust,no_run
+/// # async fn foo() {
+/// # let child: mux_client_impl::RemoteChild<'static> = unimplemented!();
+/// let stdin = child.stdin().take().unwrap();
+/// let stdout = child.stdout().take().unwrap();
+/// tokio::io::copy(&mut stdout, &mut stdin).await;
+/// # }
+/// ```
+#[derive(Debug)]
+pub struct RemoteChild<'s> {
```

Comment:
Rather than have one `RemoteChild` type per impl, how about we instead make it
```rust
struct RemoteChild<'s, C> where C: Connection {
    session: &'s Session<C>,
    state: C::RemoteChild,
}
```
And then define a `Connection` trait that we implement for both the process-based and the mux-based impl, with an associated type `RemoteChild` that holds the actual `RemoteChild` impl that we delegate to? You'll probably want a dedicated trait for `RemoteChild` (and `Command` and others) to hold the shared interface.